Load Chromium virtual buffer on state_system_busy

feerrenrut opened the pull request:
### Link to issue number:
Fixes: #13306 

### Summary of the issue:
Chromium devs wish to be able to communicate that a document is loading via [MSAA `STATE_SYSTEM_BUSY`](https://docs.microsoft.com/en-us/windows/win32/winauto/object-state-constants) [exposed via IA2](https://accessibility.linuxfoundation.org/a11yspecs/ia2/docs/html/_accessible_states_8idl.html#aafecbfb195aaf7050dc8f75a7833bd5f)

This prevents NVDA from loading a virtual buffer until the document is no longer considered busy. The definition of `STATE_SYSTEM_BUSY` is:
>The control cannot accept input at this time.

This will introduce special handling for Chromium based browsers, but will match the existing behavior for Firefox.

### Description of user facing changes
There should be no observable change in the behavior of NVDA when used with Chrome.

### Description of development approach
- Return a valid `TreeInterceptor` class from `_get_treeInterceptorClass` even when the document has `STATE_SYSTEM_BUSY`.
- Introduce a feature flag to allow disabling this behavior.
- Streamline the process of creating new feature flags. Provide reusable classes to help future flags be more consistent.

### Testing strategy:
- Ask Chromium devs to test with an NVDA try build.
- Allow users to test when the change is introduced to Chromium in the future (after this change reaches a majority of NVDA users)
- Rely on the feature flag for users to be able to disable this if there are problems.

### Known issues with pull request:
None

### Change log entries:
None

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

feerrenrut commented:
Not sure why that selection of chrome tests failed, I ran the build again an they passed.

seanbudd commented on source/gui/nvdaControls.py:
```diff
@@ -391,3 +401,125 @@ def ScrollChildIntoView(self, child: wx.Window) -> None:
+	def _getConfSpecDefaultValue(self) -> FeatureFlagValues:
+		defaultValueFromSpec = self._conf.getConfigValidation(self._confPath).default
+		if not isinstance(defaultValueFromSpec, FeatureFlag):
+			raise ValueError(f"Default spec value is not a str, but {type(defaultValueFromSpec)}")
```
Something doesn't look quite right here, is this message more accurate?

```suggestion
			raise ValueError(f"Default spec value is not a FeatureFlag, but {type(defaultValueFromSpec)}, value: {defaultValueFromSpec})
```

seanbudd commented on tests/unit/test_config.py:
```diff
@@ -0,0 +1,163 @@
+			calculatedValue=True
+		)
+
+	def test_empty_defaultsToDisabled(self):
```
Should this be

```suggestion
	def test_empty_raisesValidateError(self):
```

seanbudd reviewed (commented):
There's a few minor comments but the approach generally looks good.

General question: would it be more appropriate to add the feature flag design into a separate PR?

seanbudd commented:
I think it would be good to have another issue to convert other feature flags to use the new API. We could also consider converting other boolean flags to become feature flags.

feerrenrut commented:
> General question: would it be more appropriate to add the feature flag design into a separate PR?

I considered this, but thought it really needs to be paired with context given by usage. I could implement the feature flag "the old way", then follow up with a PR to upgrade it to the new way. That would allow for independent reverts if required.

But given the virtual buffer change is so small, I figured reverting that manually in a new commit is fairly easy.

feerrenrut commented on tests/unit/test_config.py:
```diff
@@ -0,0 +1,163 @@
+				'featureFlag(behaviorOfDefault="disabled")',
+				behaviorOfDefault="disabled"
+			),
+			'featureFlag(behaviorOfDefault="disabled", default="default")'
```
No, the doc string was wrong. The `default' arg specifies the value that should be used when the key is missing. It should always be "default", which indicates that "behaviorOfDefault" takes precedence.

Updating the PR description to clarify and add more details about this design. I meant to include this when converting from a draft.

feerrenrut replied:
Actually for this reason, it probably is best that the feature flag is it's own PR. I'm going to split this up.

feerrenrut committed:
Load vBuf in chrome when doc busy

- This is the default behavior
- Update changes file for PR #13851

feerrenrut commented on source/NVDAObjects/IAccessible/chromium.py:
```diff
@@ -64,11 +71,28 @@ def __contains__(self, obj):
+				" vbuf WILL be loaded when state of the document is busy."
+			)
+		if self.states.intersection(vbufUnavailableStates):
+			return super(Document, self).treeInterceptorClass
```
Yes, that simplifies and is equivalent because `Document` is the name of the class the method is in.

feerrenrut commented on source/NVDAObjects/IAccessible/chromium.py:
```diff
@@ -64,11 +71,28 @@ def __contains__(self, obj):
+			controlTypes.State.EDITABLE,
+		}
+		if not shouldLoadVbufOnBusyFeatureFlag:
+			log.debug(
```
No, only once per doc load. If we find that they are noisy, then we can revisit. Initially I think it will be helpful for debugging if it is very clear.

feerrenrut commented on source/config/featureFlag.py:
```diff
@@ -0,0 +1,139 @@
+	param behaviorOfDefault: Required, the default behavior of the flag, should be "enabled" or "disabled".
+	"""
+	log.debug(f"Validating feature flag: {value}, behaviorOfDefault: {behaviorOfDefault}")
+	from configobj.validate import ValidateError
```
I'll fix this.

feerrenrut commented on source/config/featureFlag.py:
```diff
@@ -0,0 +1,139 @@
+			value=specString
+		)
+	# ensure there is the expected default
+	return f'featureFlag({behaviorOfDefaultKey}="{behaviorOfDefaultVal}", default="default")'
```
No, as discussed below. The spec is being transformed to ensure that any missing featureFlag config key gets the value "default", which then refers to the "behaviorOfDefault" value.

feerrenrut commented:
See https://github.com/nvaccess/nvda/pull/13859 and commit 7ac7eaf76db45aa2a5300e44b459b30da75b1fc9 for further review actions.

seanbudd commented:
>The config will change key slightly https://github.com/nvaccess/nvda/pull/13851#discussion_r910496154 , which for alpha users means they will not have an error if using one version then the next.

This can be avoided if we don't let the first PR build complete (e.g. cancel subsequent builds) until both PRs are merged.

feerrenrut commented:
> This can be avoided if we don't let the first PR build complete (e.g. cancel subsequent builds) until both PRs are merged.

Yes. But I think it's an acceptable outcome if there are separate builds. It is confusing to have two similarly named config settings, but that's the worst of the impact (vs an error on config load and config gets reset to default)

feerrenrut committed:
Add feature flag for vBuf load on busy state (PR #13859)

See base PR #13851
Adds support for multi value feature flags from PR #13871

### Summary:
NVDA increasingly makes use of feature flags to expose experimental changes to users.
There are already several approaches already in use.
It is easy to miss use-cases for changing default behavior, potentially overriding an advanced users preferred config.

Some feature flags require more than just a Boolean enabled/disabled outcome (in addition to 'default'), E.G.:
- AllowUiaInMSWord [WHEN_NECESSARY, WHERE_SUITABLE, ALWAYS]
- AllowUiaInChromium [WHEN_NECESSARY, YES, NO]

Several approaches have been tried:
- Store an int in the config that match with enum values, as per (AllowUiaInMSWord and AllowUiaInChromium)
- Use "options" in config (several string values). Literal strings in code to compare refer to these.

These approaches can hide errors:
- Looking at the config, what are the valid values when an integer is used.
- Using string "options":
  - What are the valid options?
  - How do you find all usages?
  - Change the strategy for default?
- For both, when looking at the config or configSpec, how can the behavior of default be determined?

### Description of user facing changes
The target user of this change are developers.
This introduces constructs and utilities to streamline the approach to adding feature flags.

Adding a new feature flag:
- Most feature flags can use the BoolFlag.
- If more values are required, define a new enum class in featureFlagEnums.py (include a DEFAULT member) E.G.: MyNewFeature
- Add the entry to the configSpec.py E.G.
  shouldUseMyNewFeature = featureFlag(optionsEnumClass="MyNewFeature", behaviorOfDefault="ON_WEEKENDS")
- Add a GUI option for it, using FeatureFlagCombo
- Wrap the code for the feature with:
   config.conf["section"]["shouldUseMyNewFeature "] == MyNewFeature.ON_WEEKENDS

### Development approach
Introduction of a generic BoolFlag type (enabled/disabled) which needs to account for 3 states:
- The current NVDA default (IE user has no preference)
- User has a preference of enabled
- User has a preference of disabled

To facilitate this, one of the three values (default, enabled, disabled) is stored in the config.
Spec/config transformations are provided to streamline the support.
A featureFlag type in the spec is automatically converted to a FeatureFlag class instance.
The FeatureFlag class abstracts the consideration of "default", allowing the FeatureFlag to treated as a boolean within internal NVDA logic.

A utility GUI control is added which presents the FeatureFlag option to the user, and allows developers to give translated names to the logical values (enabled, disabled), the name for default incorporates the behaviorOfDefault (and is also translated).

Note:
- The configSpec explicitly defines what the behavior of default is.
- The enum class that backs the config is defined, allows for type / value checking.
- Existing flags have not been upgraded, these would require upgrade code for the config, this introduces unnecessary complexity.

### Known issues:
This approach bypasses the handling for defaults provided by configObj.
It was considered whether the default not written on read behavior (of the configObj library) to facilitate the third (default) state.
However:
- This would require explicit code to test for the existence of the key to know if the value returned by `conf["section"]["myKey"]` was just the default or an explicit setting.
- It also requires careful diligence to ensure that the value is not ever written accidentally.
- Code to delete the key is required whenever attempting to return to the default.
- The way validation checks are created in `configObj` makes transforming the type (to provide an abstraction) difficult. It is not possible to get access to the `default` without a similar approach used in this PR. If adopting the complexity of a spec and config transform step, then it is preferable to have explicit values for default. Explicit values make the config easier to inspect and make the approach more robust.

CyrilleB79 commented on source/config/featureFlag.py:
```diff
@@ -0,0 +1,226 @@
+		- 'behaviorOfDefault'
+		- 'optionsEnum'
+	"""
+	log.info(f"specString: {specString}, kwargs: {kwargs}")
```
Why is this log present at info level? The logged message is not self-explaining.

If it is a debug log needed only for this PR's development, it should be removed. If it has any added value to for future debug, it may be kept at debug level and maybe with a more explicit message on its goal. If there is a justification to keep it at a so high level could you please elaborate?

@feerrenrut , @seanbudd, any idea?

seanbudd replied:
This should probably be debug level or removed

feerrenrut replied:
Yes, I'll remove this. It was helpful for debugging during development.
